File: benchmark-xamarin/Benchmark.Android/MainActivity.cs
using System;
using Android.App;
using Android.Widget;
using Android.OS;
using Benchmark.Core;
using System.Linq;

namespace Benchmark.Android
{
    [Activity (Label = "Benchmark", MainLauncher = true)]
    public class MainActivity : Activity
    {

        protected override void OnCreate (Bundle bundle)
        {
            base.OnCreate (bundle);

            SetContentView (Resource.Layout.Main);

            Button button = FindViewById<Button> (Resource.Id.btnRun);
            
            button.Click += (a, b) =>
            {
                var runs = new[]{
                    new{alg= (IPrime)new AlgorithmAJavaMath(), label= FindViewById<TextView>(Resource.Id.lblAlgAJava), text = "A (java.lang.Math): "},
                    new{alg= (IPrime)new AlgorithmAMonoMath(), label= FindViewById<TextView>(Resource.Id.lblAlgAMono), text = "A (System.Math): "},
                    new{alg= (IPrime)new AlgorithmB(), label= FindViewById<TextView>(Resource.Id.lblAlgB), text = "B: "},
                };
                
                const int iterations = 100000;
                
                runs.ToList().ForEach(x =>
                {
                    var elapsed = TheBenchmark.Timed(() =>
                    {
                        for(int i=0; i<iterations; i++)
                        {
                            x.alg.IsPrime(i);
                        }
                    });
                    x.label.Text = x.text + " " + elapsed.TotalMilliseconds;
                });
            };
        }
        
        public class AlgorithmAJavaMath : IPrime
        {
            public bool IsPrime(int candidate)
            {
                if (candidate == 1) return false;
                if (candidate == 2) return true;
            
                for (int i = 2; i <= Java.Lang.Math.Ceil(Java.Lang.Math.Sqrt(candidate)); ++i)  {
                   if (candidate % i == 0)  return false;
                }
            
                return true;
            }
        }
    }
}

File: benchmark-android/benchmark/src/main/java/com/sep/benchmark/MainActivity.java
package com.sep.benchmark;

import android.support.v7.app.ActionBarActivity;
import android.os.Bundle;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.widget.Button;
import android.widget.TextView;

import java.util.GregorianCalendar;

public class MainActivity extends ActionBarActivity {

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        Button runButton = (Button)this.findViewById(R.id.btnRun);
        final TextView lblAlgorithmA = (TextView)this.findViewById(R.id.lblAlgorithmA);
        final TextView lblAlgorithmB = (TextView)this.findViewById(R.id.lblAlgorithmB);

        runButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                Prime algA = new PrimeAlgorithmA();
                Prime algB = new PrimeAlgorithmB();

                int numIterations = 100000;

                doTest(algA, numIterations, lblAlgorithmA, "A: ");
                doTest(algB, numIterations, lblAlgorithmB, "B: ");
            }
        });
    }

    public void doTest(final Prime alg, final int numIterations, TextView label, String prefix) {
        long elapsed = TimedOperation(new Runnable(){
            public void run() {
                for (int i=0; i<numIterations; i++) {
                    alg.isPrime(i);
                }
            }
        });

        label.setText(prefix + elapsed);
    }

    public long TimedOperation(Runnable toPerform) {
        long begin = GregorianCalendar.getInstance().getTimeInMillis();
        toPerform.run();
        long end = GregorianCalendar.getInstance().getTimeInMillis();

        return end - begin;
    }


    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.main, menu);
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        int id = item.getItemId();
        if (id == R.id.action_settings) {
            return true;
        }
        return super.onOptionsItemSelected(item);
    }

    interface Prime {
        boolean isPrime(int candidate);
    }

    class PrimeAlgorithmA implements Prime{
        @Override
        public boolean isPrime(int candidate) {
            if (candidate == 1) return false;
            if (candidate == 2) return true;

            for (int i = 2; i <= Math.ceil(Math.sqrt(candidate)); ++i)  {
                if (candidate % i == 0)  return false;
            }

            return true;
        }
    }

    class PrimeAlgorithmB implements Prime{
        @Override
        public boolean isPrime(int candidate) {
            if ((candidate & 1) == 0)
            {
                if (candidate == 2)
                {
                    return true;
                }
                else
                {
                    return false;
                }
            }

            for (int i = 3; (i * i) <= candidate; i += 2)
            {
                if ((candidate % i) == 0)
                {
                    return false;
                }
            }
            return candidate != 1;
        }
    }
}
